### DISCUSSION TITLE
How to fix Leading causing vertical alignment to be ignored

### DISCUSSION THREAD

**Discussion by StevenStreasick** [LibrePDF/OpenPDF#1175] [Q&A] [answered]
I am trying to set the spacing between text that is wrapped to be slightly larger. To my knowledge, this can be achieved through the leading property of a PdfPCell. However, every time that I try to add leading to my PdfPCell, it seems to suddenly ignore my vertical alignment that I assign it. Even when I set the leading to the default leading, it seems to ignore the vertical formatting. 

```
PdfPCell cell = new PdfPCell(new Phrase("Wrapped Text", font));
					
cell.setHorizontalAlignment(PdfPCell.ALIGN_LEFT);
cell.setVerticalAlignment(PdfPCell.ALIGN_MIDDLE);

cell.setLeading(0f, 1.5f);

table.addCell(cell);
doc.add(table);
```

If I try to split up the wrapped text into different chunks, I seem to encounter the same issue.

```
PdfPCell cell = new PdfPCell();

cell.setHorizontalAlignment(PdfPCell.ALIGN_LEFT);
cell.setVerticalAlignment(PdfPCell.ALIGN_MIDDLE);

cell.setLeading(0f, 1.5f);

cell.addElement(new Phrase("Wrapped", font);
cell.addElement(new Phrase("Text", font);

table.addCell(cell);
doc.add(table);
```
I am able to pad the tops and/or bottoms of the cells to try to get these to be middle-aligned, but this is a hacky solution that produces cells that are larger than desired.
```
PdfPCell cell = new PdfPCell("Wrapped Text", font);

cell.setHorizontalAlignment(PdfPCell.ALIGN_LEFT);
cell.setVerticalAlignment(PdfPCell.ALIGN_MIDDLE);
cell.setPaddingBottom(10f);

cell.setLeading(0f, 1.5f);

table.addCell(cell);
doc.add(table);
```

Does anyone know a better solution to have vertical formatting followed, while also having spacing between wrapped text lines to be followed?

Thank you in advance,
Steven

**Comment by StevenStreasick** [comment, marked as answer]
I finally found the time to investigate this issue more thoroughly. While at first, it may appear that leading causes vertical alignment to be ignored, this could be less from true. With this example, where the vertical alignment is set to ALIGN_BOTTOM, we can see that the vertical alignment is respected. 
```PdfPCell cell = new PdfPCell(new Phrase("Hello World!"));
cell.setMinimumHeight(10f);
cell.setVerticalAlignment(PdfPCell.ALIGN_BOTTOM);
cell.setHorizontalAlignment(PdfPCell.ALIGN_RIGHT);
cell.setLeading(1, 2);
randomTable.addCell(cell);
doc.add(randomTable);
```
If we look at the example below, we can start to get a better picture of what's going on.  
```
PdfPCell bigCell = new PdfPCell(new Phrase("Hello World!"));
bigCell.setMinimumHeight(200f);
bigCell.setVerticalAlignment(PdfPCell.ALIGN_TOP);
bigCell.setHorizontalAlignment(PdfPCell.ALIGN_RIGHT);
bigCell.setLeading(1, 2);
randomTable.addCell(bigCell);
doc.add(randomTable);
```
Upon analyzing the source code of OpenPDF, particularly regarding the handling of leading for PdfPCells and ColumnTexts, it becomes apparent that the current implementation deviates from the standard definition of leading as described by Adobe. According to Adobe, "leading" refers to the vertical space between lines of type.

In the current implementation, leading is applied between the margin and the first line that appears for PdfPCells and ColumnTexts. This is counterintuitive to Adobe's definition, where leading is defined as affecting the space between successive lines of text.

To address this issue, a straightforward solution would be to introduce a boolean variable in the 'go(boolean)' method which could effectively distinguish whether the current cell is the first in sequence. This addition would enable precise control over when leading should be applied, thereby resolving the above discrepancy.

Edit: I found a built in solution to this problem which acts nearly the same as my proposed solution. Column Texts and in turn PdfPCells have a function called 'setUseAscender', which determines if the leading will apply to the first line or not. I find it an interesting choice to have this default to false, but it is nice to have a proper fix.

**Reply by MMKe** [reply]
Thank you for your sharing.
